### ISSUE TITLE
Add PrimMonad instance for CatchT

### ISSUE THREAD

**Issue by treeowl** [haskell/primitive#218]
`exceptions` defines a `CatchT` type that can be given a `PrimMonad` instance. Should we accept the dependency to add it?

**Comment by andrewthad** [comment]
I wouldn't like a dependency on `exceptions`. I would prefer that `exceptions` depend on `primitive`.

**Comment by chessai** [comment]
Agreed about the dependency thing. also the docs of CatchT mention that `m` should not unify with `IO`. I guess `ST s` might be fine?

**Comment by treeowl** [comment]
The problem is that `primitive` is not a terribly small dependency. Some people may be concerned about build time. I would probably rather move `PrimMonad` and `PrimBase` to the `base` library (in `GHC.StateTransformers` or something). The only tricky thing about that is that we'll have to convince Ross Paterson to put the necessary instances in `transformers`. An easier option would be to spin `Control.Monad.Primitive` off into its very own package, say `prim-monad`, that depends solely on `base` and is too small for (almost) anyone to worry about. There are a couple other pieces of this package that I think would be better off n their own packages: `PrimUnlifted` and `Prim`. Any time I see a package that could easily be split into several different ones, I tend to think it should be. This package, I believe, would split much more cleanly than some others that actually have been split.

**Comment by andrewthad** [comment]
I don't think of `primitive` as as big depedency. But, I could get behind the idea of moving `PrimMonad`, `PrimBase`, `PrimUnlifted`, and `Prim` into either some kind of `primitive-class` package or into `base`. I generally have a favorable view of bringing the most fundamental types and typeclasses into `base` (leaving functions in their own packages). `PrimUnlifted` certainly meets this standard, although it would be nice to remove the legacy methods that use `unsafeCoerce#` from it before enshrining it in `base`. I feel that the other three typeclasses are fundamental enough to go in `base`, but there's at least a little room for argument on those.

**Comment by treeowl** [comment]
I'd much rather *not* lump all the classes into one package, unless that is `base`. `PrimMonad`, `PrimUnlifted`, and `Prim` are three completely independent concepts which can be implemented completely independently. They're only tied together "a level up" by functions that use two of those.

**Comment by andrewthad** [comment]
That's fair. Since `PrimUnlifted` should not be moved until `primitive`'s support for GHC 8.0 is dropped, it's probably worth checking on the libraries mailing list to see what people think about moving `PrimBase` and `PrimMonad`. But just the typeclasses. Not all their helper functions. That would solve the problem at hand.

**Comment by treeowl** [comment]
I can ask. Remember, though, that Paterson can effectively veto moving those two to `base`, so he must be convinced as well.

**Comment by treeowl** [comment]
The upgrade path is also non-trivial, so we'd need a plan.

**Comment by andrewthad** [comment]
The upgrade path I imagined was:

- Add a module named `Control.Monad.Primitive.Class` to `base` that just has the two typeclasses.
- In `Control.Monad.Primitive`, which still lives in `primitive`, use a little CPP to reexport those typeclasses if we are building against an newer base and define them if we are building against a newer base.

**Comment by treeowl** [comment]
@cartazio, can you consider my reasoning for splitting the package up rather than just rejecting out of hand? Users would only need to depend on `primitive` regardless, as it would provide all the modules from the sub-packages.

**Comment by cartazio** [comment]
i'm all for splitting out the unlifted and array kits into sub systems, so i aggree, just which is where, 

i think more like primitive as the simple one, then -unlifted and -primarray  and perhaps a -fancy overlay for experimental / exploratory stuff that needed be as long lived / stable

**Comment by treeowl** [comment]
@cartazio, I see. I had been operating on the assumption that we'd want as little breakage as possible. Moving things altogether out of `primitive` (so users have to add dependencies or switch to `primitive-batteries` or whatever) is major breakage. Breaking out pieces and making `primitive` depend on them avoids any breakage at all.

**Comment by cartazio** [comment]
very little code depends on unlifted and primarray, and theyre more complicated / less used by ecosystem, so i'm inclined to make it easier to workshop / iterate on those

**Comment by cartazio** [comment]
(so i'm not terribly concerned about user impact from moving those out, theyre both pretty recent)

**Comment by treeowl** [comment]
> very little code depends on unlifted and primarray, and theyre more complicated / less used by ecosystem, so i'm inclined to make it easier to workshop / iterate on those

That  may be, but `Prim` and the `ByteArray` and `Addr` operations it supports have been around for some time. `Prim`, however, has nothing at all to do with `PrimMonad`.

**Comment by cartazio** [comment]
i'm working through some of the implications of different parts,

im definitely comfortable moving unlifted tools out into its own thing, and def not 100% on the prim/byte array bits.

I've struggled to find any widely used/depended upon code in the wild that uses address. aside from the one gross spot in vector, have you seen any?

@treeowl it occurs to me, maybe we should create a "modern-pointer-experiments" package so we can workshop out some various modern style/improved pointer apis for haskell?

**Comment by andrewthad** [comment]
I make pretty heavy use of `Addr` in some of the software I work on for my job and in several unpublished libraries on my github. I am likely the main user of this API.

**Comment by cartazio** [comment]
Definitely!

Yeah, I spoke with chessai about that and it seems like it’d be easy to fix
on your side, or at least those apis would work just as correctly with say
‘Ptr a’, moral aesthetics / quality issues with say the storable pointer
api (we should experiment with better pointer interfaces)

**Comment by cartazio** [comment]
*aesthetic issues in the Address vs  Ptr a vs current pointer apis could be
made better aside

**Comment by chessai** [comment]
That is not a shared opinion.

Also I think that this issue has derailed a bit. The splitting off of
typeclasses into base or other packages should probably be its own issue.
